src/errors.rs

```diff
@@ -1,33 +1,126 @@
-use core::fmt::Display;
-use failure::Context;
+use core::fmt;
+use std::borrow::Cow;
+use std::io;
 
-#[derive(Debug, Fail)]
-pub enum ErrorKind {
-    #[fail(display = "json rpc error")]
-    JsonRpc,
-    #[fail(display = "core rpc error")]
-    CoreRpc,
-    #[fail(display = "mnemonic error")]
-    Mnemonic,
+use bitcoin::consensus::encode;
+use bitcoin::util::bip32;
+use bitcoincore_rpc;
+use failure;
+use hex;
+use secp256k1;
+use serde_json;
+use url;
+
+pub const GDK_ERROR_ID_UNKNOWN: &'static str = "id_unknown";
+
+const CORE_INSUFFICIENT_FUNDS: i32 = -1;
+
+#[derive(Debug)]
+pub enum Error {
+    // First we specify exact errors that map GDK errors.
+    InsufficientFunds,
+
+    /// Special case error to explicitly specify the GDK error
+    /// code without having an error type for it.
+    /// This should be avoided.
+    Gdk(&'static str),
+
+    // And then all other errors that we can't convert to GDK codes.
+    Bip32(bip32::Error),
+    Bip39(failure::Error),
+    BitcoinEncode(encode::Error),
+    BitcoinRpc(bitcoincore_rpc::Error),
+    Hashes(bitcoin_hashes::Error),
+    Hex(hex::FromHexError),
+    Io(io::Error),
+    Json(serde_json::Error),
+    Secp256k1(secp256k1::Error),
+    Url(url::ParseError),
+
+    /// Custom error with message.
+    Other(Cow<'static, str>),
+}
+
+impl Error {
+    /// Convert the error to a GDK-compatible code.
+    pub fn to_gdk_code(&self) -> &'static str {
+        // Unhandles error codes:
+        // id_no_utxos_found
+        // id_no_recipients
+        // id_no_amount_specified
+        // id_fee_rate_is_below_minimum
+        // id_invalid_replacement_fee_rate
+        // id_send_all_requires_a_single_output
+        match *self {
+            Error::InsufficientFunds => "id_insufficient_funds",
```

Comment:
The explicit error id special-case is missing here (`Error::Gdk(err) => err`)

Reply:
I removed it.

Reply:
Alternatively, the special case could just be removed in favor of always creating specialized error variants.